### DISCUSSION TITLE
MIP Provider Authorization (AKA CAIP-25)

### DISCUSSION THREAD

**Discussion by shanejonas** [MetaMask/metamask-improvement-proposals#33] [MIPs]
This is the discussion thread for Provider Authorization (AKA CAIP-25).

It is currently in draft state you can view it here: https://github.com/MetaMask/metamask-improvement-proposals/pull/31

**Comment by shanejonas** [comment]
My thoughts after reviewing the MIP:

### Restricting Methods Over New Transport

- **Critical Point**: Hesitation on prohibiting certain methods. This could potentially hinder the adoption of the multichain API by significant libraries (e.g., WalletKit) that prioritize backward compatibility for their user base.
- **Sessions Concern**: The concept of sessions remains unclear. Allowing multiple sessions for the same domain could introduce complexity. Is there a solid justification for this approach?
- **Multichain API Accessibility**: The decision to restrict multichain API access via the existing Ethereum API is questionable. As Jiexi highlighted, limiting access contrasts sharply with the idea of enhancing functionality by merely adding new methods. This decision could impact usability and integration efforts negatively.

### Optional Scopes

- **Bug Issue**: The necessity to make all scopes optional due to a bug is concerning. Before fully committing to CAIP-25, a thorough investigation into this issue is essential, especially given the indication of deprecated elements. The reliance on potentially unstable or changing standards could introduce risks to implementation and future compatibility.
    - **Reference**: [WalletConnect's CAIP-25 Implementation Guidance](https://walletconnect.com/blog/caip-25-implementation-guidance-migrating-to-empty-undefined-required-namespaces). This article mentions the workaround for the bug, implying a need for caution and further analysis before adoption.

**Comment by bumblefudge** [comment]
Apologies if I'm stumbling in off the street without context, but I thought it might be helpful to chip in my thoughts and at least volunteer to answer questions about, if nothing else, how WC is using CAIP-25.

* backwards compatibility: I think I agree? is there further discussion about this somewhere someone could link me to? If deprecating some methods or making them slower is an option, I would advocate for that over breakage, as the biggest problem with CAIP-25 has been the upgrade path from status quo!
* sessions concern: I am a little unclear myself-- it was my understanding that CAIP-25 establishing multiple scopes within *one session* between application and provider, and both sides persist a session object containing all the CAIP-25 information and update it over time in a progressive-authorization use-case; I think the handshake MIP refers to this as "a concurrent session", meaning one session that concurrently maintains multiple providers or RPC connections in parallel, *sandboxed from each other to the extent of the wallet's ability*, rather than multiple sessions (in the browser API sense of the word).  None of that is in the document, though, so maybe I'm [erroneously] extrapolating in outside context? here as well links to where session mental models are defined would help me avoid wasting time discussing what's out of scope :D
* the wallet connect blog post doesn't mention any bugs, if anything it was a documentation/developer-relations glitch in that developers didn't realize the failure mode of marking all requested scopes as "mandatory" because the examples in early drafts of the developer docs didn't match the accompanying prose.  I definitely agree that investigating both UX and DX makes sense for a big change, with regards to both the deprecation issue and the shift to a complex negotiation protocol for permissions/authZ, but really the way to avoid the pain wallet connect experienced rolling this out is to lavish attention on the documentation and, most importantly of all, the examples and defaults and templates contained therein 😎

**Comment by adonesky1** [comment]
> Critical Point: Hesitation on prohibiting certain methods. This could potentially hinder the adoption of the multichain API by significant libraries (e.g., WalletKit) that prioritize backward compatibility for their user base.

This will be a new API so the backwards compatibility issues are all front loaded. 1) this will require wrapping methods in CAIP-27 request wrapper, 2) this will be served through a new transport layer. Part of the point is to consolidate all of these breaking changes into one batch so we don't have to introduce big breaking changes (new API!) and then some months later be like oh heres some more huge breaking changes (new transport layer!) and then more breaking changes (we're deprecating methods x,y and z)

When we say prohibiting methods are you referring to?

```
 wallet_switchEthereumChain method WILL NOT be carried over to the Multichain API as chain switching coordination between dApp and wallet will no longer be necessary
eth_sign`, `eth_signTypedData`, `eth_signTypedData_v3` methods WILL NOT be included in the Multichain API as these signature methods have been superseded by the more commonly used `personal_sign` and `eth_signTypedData_v4` methods
```

- We’re removing `wallet_switchEthereumChain` because it won’t be relevant for this API
- `eth_sign` is already deprecated in our existing API
- I suppose there is some argument that not carrying over `eth_signTypedData`, `eth_signTypedData_v3` could be a problem for some dapps but really its probably time to rip off that bandaid?

> Sessions Concern: The concept of sessions remains unclear. Allowing multiple sessions for the same domain could introduce complexity. Is there a solid justification for this approach?

We needn’t treat sessions in the way that WalletConnect might be treating them. First of all the spec seems to indicate that `sessionProperties` are optional: ‘A third object is the sessionProperties object, which also MUST contain 1 or more properties if present.’

Second we have been thinking of it more for this purpose described in the CAIP-25 spec: “In addition to making properties of the negotiated session itself explicit, they can also annotate, support, or extend the negotiation of scope proposals (e.g., providing information about unfamiliar scopes or which accounts to expose to each).”  These could allow us some flexibility to evolve our use of CAIP-25 and adapt it further for use with say Snaps for instance.  

> Bug Issue: The necessity to make all scopes optional due to a bug is concerning. Before fully committing to CAIP-25, a thorough investigation into this issue is essential, especially given the indication of deprecated elements. The reliance on potentially unstable or changing standards could introduce risks to implementation and future compatibility.

I don’t think this is a bug. It was more of a UX oversight that we’ve already accounted for in our multichain UX planning. Basically we can continue to follow the spec exactly but will actively discourage consumers (and libraries especially) from using requiredScopes for the same reasons outlined in the WalletConnect post

**Comment by vandan** [comment]
Re: typed data signature methods
I consulted with @bschorchit about the signature methods we plan to exclude from the new API and it sounds like there would be minimal anticipated impact as most dapps have now moved on to using signTypedData_v4.

The "Multichain API Accessibility" topic seems to be getting at concerns about barriers to developer adoption, which are important. Of course, the API has little value without that. At the same time, the multichain API inherently introduces new paradigms that we would need developers to adopt anyways, right? My understanding was that these perceived barriers could be mitigated by dapps or third-party libs like ethers.js, wagmi, MM SDK etc. if they/we provide "adapters" to allow some or all of the existing methods to be mapped to corresponding multichain APIs (or handled as no-ops, etc.). Are there methods we consider as not being mappable from the current API to the proposed multichain API?

**Comment by adonesky1** [comment]
@bumblefudge and @hmalik88 wanted to loop back here on a few topics that our team would like to discuss before moving forward with implementation:

**1.** Not entirely clear why we need sessionIdentifiers for extension wallets like MetaMask as discussed in CAIP-25 and [CAIP-171](https://chainagnostic.org/CAIPs/caip-171). CAIP-171 refers to closable sessions? What doe closing a session for dapp connections with MetaMask mean? Since its not intermediated by a socket connection as with WalletConnect?
   **1A.** Since in our case "sessions" really translate to long lived permissions we would probably want some equivalent of [`wallet_revokePermissions`](https://github.com/MetaMask/metamask-improvement-proposals/blob/main/MIPs/mip-2.md) that terminates "sessions" between dapp and wallet? `provider_unauthorize`?
**2.** We think CAIP-25 needs to be more specific about how responses should look when say a user rejects some scopes or the users wallet doesn’t have the capability for some proposed scopes.

There maybe some others that I'm forgetting ATM but we can start by discussing these at least!

**Reply by bumblefudge** [reply]
1. I can't put words in Olaf's or Hassan's mouth, but it sounds a little to me like 1A kind of answers 1?  The idea was that wherever that session can be closed, both parties need to know: if the provider closes it, no need for an RPC, if wallet or caller close it, they need to notify one another-- by calling some equivalent of `wallet_revokePermissions` on that session Identifier or specific contents inside of the session object it identifies, right?  It's exactly because the models differ in how that state is maintained that an identifier across both needed to be standardized in advance 😄 

2. Can you say more? The opacity of which scopes or features were rejected for inability to support versus which were rejected for lack of permissions was deliberate, to minimize fingerprinting risk.  the error codes are given with the double-caveat that they are normatively optional and not recommended for untrusted callers. This was the result of conversations with dan F and kyle dH from brave browser, if memory serves-- happy to advise further, tho, it might still require some finesse to mitigate this strategy on the implementation side?

**Reply by adonesky1** [reply]
> if wallet or caller close it, they need to notify one another-- by calling some equivalent of wallet_revokePermissions on that session Identifier or specific contents inside of the session object it identifies, right?

Yeah exactly. So its a bit ambiguous in CAIP-25 how a caller can signal the end of a session to the wallet without signaling an intent to start a new one: _"If a respondent (e.g. a wallet) needs to initiate a new session, whether due to user input, security policy, or session expiry reasons, it can simply generate a new session identifier to signal this notification to the calling provider; if a caller needs to initiate a new session, it can do so by sending a new request without a sessionIdentifier"_

I suppose you could just send a CAIP-25 request with no scopes? 

This isn't a mission critical feature but one that we recently introduced via [MIP-2](https://github.com/MetaMask/metamask-improvement-proposals/blob/main/MIPs/mip-2.md) (`wallet_revokePermissions`) so would ideally replicate in some form in this API as well.

> The opacity of which scopes or features were rejected for inability to support versus which were rejected for lack of permissions was deliberate, to minimize fingerprinting risk.

Yeah I don't think I thought this through enough before asking. If the user rejects some optional scopes no need to say anything in the response about that other than they are absent from the response sessionScopes. And if anything is rejected or can't be provided in requiredScopes we ought to give an opaque failure to avoid fingerprinting

**Reply by hmalik88** [reply]
Replied to these concerns in a DM, can surface to here as well for transparency. I think it might be worth loosening the requirements around `sessionId`. I'll get in touch to set up a meeting to discuss @bumblefudge
